advertise/addr.go

```diff
@@ -5,99 +5,164 @@ package advertise
 		if err != nil {
 			err = fmt.Errorf("interface %q: %w", ifaceName, err)
 			errs = multierror.Append(errs, err)
 			continue
 		}
 
-		addrs, err := iface.Addrs()
-		if err != nil {
-			err = fmt.Errorf("interface %q addrs: %w", ifaceName, err)
-			errs = multierror.Append(errs, err)
-			continue
-		} else if len(addrs) <= 0 {
-			err = fmt.Errorf("interface %q has no addresses", ifaceName)
-			errs = multierror.Append(errs, err)
+		canditate := filterBestIP(addrs)
+		if !canditate.IsValid() {
 			continue
 		}
 
-		foundAddr := findSuitableIP(addrs)
-		if foundAddr == nil {
-			err = fmt.Errorf("interface %q has no suitable addresses", ifaceName)
-			errs = multierror.Append(errs, err)
-			continue
+		if canditate.Is4() && !canditate.IsLinkLocalUnicast() {
+			// Best address possible, we can return early.
+			return canditate.String(), nil
 		}
 
-		if !IsAutomaticPrivateIP(foundAddr) {
-			return foundAddr, nil
-		} else if privateIP == nil {
-			privateIP = foundAddr
+		bestIP = filterBestIP([]netip.Addr{canditate, bestIP})
+	}
+	if !bestIP.IsValid() {
+		if errs != nil {
+			return "", errors.Wrapf(errs, "no useable address found for interfaces %v", interfaces)
+		} else {
+			return "", fmt.Errorf("no useable address found for interfaces %v", interfaces)
 		}
 	}
+	return bestIP.String(), nil
+}
+
+// getInterfaceAddresses is the standard approach to collecting []net.Addr from a network interface by name.
+func getInterfaceAddresses(name string) ([]netip.Addr, error) {
+	inf, err := net.InterfaceByName(name)
+	if err != nil {
+		return nil, err
+	}
+
+	addrs, err := inf.Addrs()
+	if err != nil {
+		return nil, err
+	}
 
-	if privateIP == nil {
-		return nil, errs.ErrorOrNil()
+	// Using netip.Addr to allow for easier and consistent address parsing.
+	// Without this, the net.ParseCIDR() that we might like to use in a test does
+	// not have the same net.Addr implementation that we get from calling
+	// interface.Addrs() as above.  Here we normalize on netip.Addr.
+	netaddrs := make([]netip.Addr, len(addrs))
+	for i, a := range addrs {
+		prefix, err := netip.ParsePrefix(a.String())
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to parse netip.Prefix")
+		}
+		netaddrs[i] = prefix.Addr()
 	}
-	return privateIP, nil
+
+	return netaddrs, nil
 }
 
-// findSuitableIP searches addrs for the first IPv4 address. IPv4 addresses
-// used for APIPA will be ignored if possible.
-//
-// Returns nil if no suitable addresses were found.
-func findSuitableIP(addrs []net.Addr) net.IP {
-	var privateIP net.IP
+// filterBestIP returns an opinionated "best" address from a list of addresses.
+// The ordering is the following:
+// - IPv4 valid and not link-local unicast
+// - IPv6 valid and not link-local unicast
+// - IPv4 valid and link-local unicast
+// - IPv6 valid and link-local unicast
+// If none of the above are found, an invalid address is returned.
+// Loopback addresses are never selected.
```

Comment:
Since the [public documentation](https://pkg.go.dev/github.com/grafana/ckit/advertise) of the package will only include docstrings for exported methods, I'd add these to `FirstAddress`.